packages/typespec-client-generator-core/src/validate.ts

```diff
@@ -1,12 +1,177 @@
-import { Program } from "@typespec/compiler";
-import { createTCGCContext } from "./internal-utils.js";
-import { getAllModelsWithDiagnostics } from "./types.js";
+import {
+  Enum,
+  EnumMember,
+  Interface,
+  listServices,
+  Model,
+  ModelProperty,
+  Namespace,
+  Operation,
+  Program,
+  Scalar,
+  Type,
+  Union,
+  UnionVariant,
+} from "@typespec/compiler";
+import { DuplicateTracker } from "@typespec/compiler/utils";
+import { getClientNameOverride } from "./decorators.js";
+import { AllScopes, clientNameKey, createTCGCContext, TCGCContext } from "./internal-utils.js";
+import { reportDiagnostic } from "./lib.js";
 
 export function $onValidate(program: Program) {
-  // Pass along any diagnostics that might be returned from the HTTP library
   const tcgcContext = createTCGCContext(program);
-  const [_, diagnostics] = getAllModelsWithDiagnostics(tcgcContext);
-  if (diagnostics.length > 0) {
-    program.reportDiagnostics(diagnostics);
+  const languageScopes = getDefinedLanguageScopes(program);
+  for (const scope of languageScopes) {
+    validateClientNames(program, tcgcContext, scope);
+  }
+}
+
+function getDefinedLanguageScopes(program: Program): Set<string | symbol> {
+  const languageScopes = new Set<string | symbol>();
+  for (const value of program.stateMap(clientNameKey).values()) {
+    if (value[AllScopes]) {
+      languageScopes.add(AllScopes);
+    }
+    for (const languageScope of Object.keys(value)) {
+      languageScopes.add(languageScope);
+    }
+  }
+  return languageScopes;
+}
+
+function validateClientNames(program: Program, tcgcContext: TCGCContext, scope: string | symbol) {
+  const services = listServices(program);
```

Comment:
Feel like you should just be able to omit all here and just always start at the global namespace as you recurrsively look into down

Reply:
You are right, that would just work

Reply:
I still see this, can't you just call 
```
validateClientNamesPerNamespace(program, tcgcContext, scope, program.getGlobalNamespaceType());
```

Reply:
> I still see this, can't you just call
> 
> ```ts
> validateClientNamesPerNamespace(program, tcgcContext, scope, program.getGlobalNamespaceType());
> ```

Think this doesn't handle multiple services.

Reply:
This will traverse every namespaces, if not you are missing a recursion anyway to look at sub namespace or the service

Reply:
I was wondering whether or not 1 namespace is guaranteed to have 1 namespace.
If so, I am fine with this change.

Reply:
changed it back, please let me know if anything is left within this PR.

Reply:
having a service or not feels irrelevant of the use of `@clientName`. Everything is under the global namespace in the type graph

Reply:
right, this is traversing everything within a namespace not a service.